File: app/src/main/java/com/cyl/musiclake/ui/music/list/presenter/PlaylistDetailPresenter.java
package com.cyl.musiclake.ui.music.list.presenter;

import com.cyl.musiclake.RxBus;
import com.cyl.musiclake.base.BasePresenter;
import com.cyl.musiclake.bean.Music;
import com.cyl.musiclake.bean.Playlist;
import com.cyl.musiclake.data.AppRepository;
import com.cyl.musiclake.event.PlaylistEvent;
import com.cyl.musiclake.musicapi.MusicApiServiceImpl;
import com.cyl.musiclake.ui.music.list.contract.PlaylistDetailContract;
import com.cyl.musiclake.utils.CoverLoader;
import com.cyl.musiclake.utils.LogUtil;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;


/**
 * Created by yonglong on 2018/1/7.
 */

public class PlaylistDetailPresenter extends BasePresenter<PlaylistDetailContract.View> implements PlaylistDetailContract.Presenter {
    private static final String TAG = "PlaylistDetailPresenter";

    @Inject
    public PlaylistDetailPresenter() {
    }
    @Override
    public void loadPlaylistSongs(String playlistID) {
        MusicApiServiceImpl.getMusicList(playlistID)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<Music>>() {
                    @Override
                    public void onSubscribe(Disposable disposable) {

                    }

                    @Override
                    public void onNext(List<Music> musicInfos) {
                        mView.showPlaylistSongs(musicInfos);
                    }

                    @Override
                    public void onError(Throwable throwable) {
                        LogUtil.e(TAG, throwable.getMessage());
                    }

                    @Override
                    public void onComplete() {

                    }
                });
//        AppRepository.getPlaylistSongsRepository(mContext, playlistID)
//                .subscribeOn(Schedulers.io())
//                .observeOn(AndroidSchedulers.mainThread())
//                .subscribe(new Observer<List<Music>>() {
//
//                    @Override
//                    public void onSubscribe(Disposable d) {
//
//                    }
//
//                    @Override
//                    public void onNext(List<Music> musicList) {
//                        if (musicList.size() == 0) {
//                            mView.showEmptyView();
//                        }
//                        mView.showPlaylistSongs(musicList);
//                    }
//
//                    @Override
//                    public void onError(Throwable e) {
//                        mView.hideLoading();
//                        e.printStackTrace();
//                    }
//
//                    @Override
//                    public void onComplete() {
//                        mView.hideLoading();
//                    }
//                });
    }

    @Override
    public void loadPlaylistArt(String playlistID) {
        AppRepository.getPlaylistSongsRepository(mView.getContext(), playlistID)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<Music>>() {

                    @Override
                    public void onSubscribe(Disposable d) {
                    }

                    @Override
                    public void onNext(List<Music> musicList) {
                        String url = null;
                        for (int i = 0; i < musicList.size(); i++) {
                            url = musicList.get(i).getCoverUri();
                            if (url != null)
                                break;
                        }
                        CoverLoader.loadBitmap(mView.getContext(), url, bitmap -> {
                            mView.showPlaylistArt(bitmap);
                        });
                    }

                    @Override
                    public void onError(Throwable e) {
                        e.printStackTrace();
                    }

                    @Override
                    public void onComplete() {
                    }
                });
    }

    @Override
    public void deletePlaylist(Playlist playlist) {
        MusicApiServiceImpl.deletePlaylist(playlist.getId())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Integer>() {
                    @Override
                    public void onSubscribe(Disposable disposable) {

                    }

                    @Override
                    public void onNext(Integer integer) {
                        if (integer == 1) {
                            mView.success(1);
                            RxBus.getInstance().post(new PlaylistEvent());
                        }
                    }

                    @Override
                    public void onError(Throwable throwable) {

                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    @Override
    public void renamePlaylist(Playlist playlist, String title) {
        MusicApiServiceImpl.renamePlaylist(playlist.getId(), title)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<String>() {
                    @Override
                    public void onSubscribe(Disposable disposable) {

                    }

                    @Override
                    public void onNext(String status) {
                        LogUtil.e(TAG, status);
                        if (status.equals("true")) {
                            mView.success(1);
                            RxBus.getInstance().post(new PlaylistEvent());
                        }
                    }

                    @Override
                    public void onError(Throwable throwable) {
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    @Override
    public void uncollectMusic(String pid, int position, Music music) {
        MusicApiServiceImpl.uncollectMusic(pid, music)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<String>() {
                    @Override
                    public void onSubscribe(Disposable disposable) {

                    }

                    @Override
                    public void onNext(String status) {
                        LogUtil.e(TAG, status);
                        if (status.equals("true")) {
                            mView.removeMusic(position);
//                            RxBus.getInstance().post(new PlaylistEvent());
                        }
                    }

                    @Override
                    public void onError(Throwable throwable) {
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }
}

File: app/src/main/java/com/cyl/musiclake/ui/music/playlist/detail/PlaylistDetailPresenter.kt
package com.cyl.musiclake.ui.music.playlist.detail

import com.cyl.musiclake.api.music.MusicApiServiceImpl
import com.cyl.musiclake.api.music.baidu.BaiduApiServiceImpl
import com.cyl.musiclake.api.music.netease.NeteaseApiServiceImpl
import com.cyl.musiclake.api.net.ApiManager
import com.cyl.musiclake.api.net.RequestCallBack
import com.cyl.musiclake.api.playlist.PlaylistApiServiceImpl
import com.cyl.musiclake.bean.Album
import com.cyl.musiclake.bean.Artist
import com.cyl.musiclake.bean.Music
import com.cyl.musiclake.bean.Playlist
import com.cyl.musiclake.data.PlaylistLoader
import com.cyl.musiclake.data.SongLoader
import com.cyl.musiclake.common.Constants
import com.cyl.musiclake.event.MyPlaylistEvent
import com.cyl.musiclake.ui.base.BasePresenter
import com.cyl.musiclake.utils.LogUtil
import com.cyl.musiclake.utils.ToastUtils
import org.greenrobot.eventbus.EventBus
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread
import javax.inject.Inject


/**
 * Created by yonglong on 2018/1/7.
 */

class PlaylistDetailPresenter @Inject
constructor() : BasePresenter<PlaylistDetailContract.View>(), PlaylistDetailContract.Presenter {

    /**
     * 加载歌手歌曲列表
     */
    override fun loadArtistSongs(artist: Artist) {
        if (artist.type == null || artist.type == Constants.LOCAL) {
            doAsync {
                val data = SongLoader.getSongsForArtist(artist.name)
                uiThread {
                    mView.showPlaylistSongs(data)
                }
            }
            return
        } else if (artist.type == Constants.BAIDU) {
            val observable = BaiduApiServiceImpl.getArtistSongList(artist.artistId.toString(), 0)
            ApiManager.request(observable, object : RequestCallBack<Artist> {
                override fun success(result: Artist) {
                    mView?.showPlaylistSongs(result.songs)
                }

                override fun error(msg: String) {
                    LogUtil.e(TAG, msg)
                    mView?.showError(msg, true)
                    ToastUtils.show(msg)
                }
            })
            return
        }
        val observable = MusicApiServiceImpl.getArtistSongs(artist.type!!, artist.artistId.toString(), 50, 0)
        ApiManager.request(observable, object : RequestCallBack<Artist> {
            override fun success(result: Artist) {
                val musicLists = result.songs
                val iterator = musicLists.iterator()
                while (iterator.hasNext()) {
                    val temp = iterator.next()
                    if (temp.isCp) {
                        //list.remove(temp);// 出现java.util.ConcurrentModificationException
                        iterator.remove()// 推荐使用
                    }
                }
                mView?.showPlaylistSongs(musicLists)
            }

            override fun error(msg: String) {
                LogUtil.e(TAG, msg)
                mView?.showError(msg, true)
                ToastUtils.show(msg)
            }
        })
    }

    /**
     * 加载专辑歌曲列表
     */
    override fun loadAlbumSongs(album: Album) {
        if (album.type == null || album.type == Constants.LOCAL) {
            doAsync {
                val data = SongLoader.getSongsForAlbum(album.name)
                uiThread {
                    mView.showPlaylistSongs(data)
                }
            }
            return
        } else if (album.albumId == null) {
            mView?.showPlaylistSongs(null)
            return
        } else if (album.type == Constants.BAIDU) {
            val observable = BaiduApiServiceImpl.getAlbumSongList(album.albumId.toString())
            ApiManager.request(observable, object : RequestCallBack<Album> {
                override fun success(result: Album) {
                    mView?.showPlaylistSongs(result.songs)
                }

                override fun error(msg: String) {
                    LogUtil.e(TAG, msg)
                    mView?.showError(msg, true)
                    ToastUtils.show(msg)
                }
            })
            return
        }
        val observable = MusicApiServiceImpl.getAlbumSongs(album.type.toString(), album.albumId.toString())
        ApiManager.request(observable, object : RequestCallBack<Album> {
            override fun success(result: Album) {
                result.name?.let { mView?.showTitle(it) }
                result.cover?.let { mView?.showCover(it) }
                result.info?.let { mView?.showDescInfo(it) }
                mView?.showPlaylistSongs(result.songs)
            }

            override fun error(msg: String) {
                LogUtil.e(TAG, msg)
                mView?.showError(msg, true)
                ToastUtils.show(msg)
            }
        })
    }

    /**
     * 加载歌单歌曲列表
     */
    override fun loadPlaylistSongs(playlist: Playlist) {
        when (playlist.type) {
            Constants.PLAYLIST_LOCAL_ID,
            Constants.PLAYLIST_HISTORY_ID,
            Constants.PLAYLIST_LOVE_ID,
            Constants.PLAYLIST_QUEUE_ID -> doAsync {
                val data = playlist.pid?.let { PlaylistLoader.getMusicForPlaylist(it, playlist.order) }
                uiThread {
                    if (data != null && data.isNotEmpty()) {
                        mView?.showPlaylistSongs(data)
                    } else {
                        mView?.showEmptyState()
                    }
                }
            }
            Constants.PLAYLIST_BD_ID -> {
                ApiManager.request(BaiduApiServiceImpl.getRadioChannelInfo(playlist), object : RequestCallBack<Playlist> {
                    override fun error(msg: String?) {
                        mView?.showError(msg, true)
                    }

                    override fun success(result: Playlist?) {
                        result?.let {
                            if (it.musicList.isNotEmpty()) {
                                mView?.showPlaylistSongs(it.musicList)
                            } else {
                                mView?.showEmptyState()
                            }
                        }
                    }

                })

            }
            Constants.PLAYLIST_WY_ID -> {
                ApiManager.request(playlist.pid?.let { NeteaseApiServiceImpl.getPlaylistDetail(it) }, object : RequestCallBack<Playlist> {
                    override fun error(msg: String?) {
                        mView?.showError(msg, true)
                    }

                    override fun success(result: Playlist?) {
                        result?.let {
                            if (it.musicList.isNotEmpty()) {
                                mView?.showPlaylistSongs(it.musicList)
                            } else {
                                mView?.showEmptyState()
                            }
                        }
                    }

                })

            }
            Constants.PLAYLIST_WY_RECOMMEND_ID -> {
                loadRecommendSongs()
            }
            else -> ApiManager.request(playlist.pid?.let { PlaylistApiServiceImpl.getMusicList(it) }, object : RequestCallBack<MutableList<Music>> {
                override fun success(result: MutableList<Music>) {
                    mView?.showPlaylistSongs(result)
                }

                override fun error(msg: String) {
                    LogUtil.e(TAG, msg)
                    mView?.showError(msg, true)
                    ToastUtils.show(msg)
                }
            })
        }
    }

    /**
     * 加载每日推荐歌曲（需登录）
     */
    private fun loadRecommendSongs() {
        val observable = NeteaseApiServiceImpl.recommendSongs()
        ApiManager.request(observable, object : RequestCallBack<MutableList<Music>> {
            override fun success(result: MutableList<Music>) {
                mView?.showPlaylistSongs(result)
            }

            override fun error(msg: String) {
                if (msg == "需要登录") {
                    mView?.showErrorTips("需要绑定网易云音乐账号", hasTry = true)
                } else {
                    mView?.showErrorTips(msg, hasTry = true)
                }
            }
        })
    }

    /**
     * 删除歌单
     */
    override fun deletePlaylist(playlist: Playlist) {

    }

    /**
     * 重命名歌单
     */
    override fun renamePlaylist(playlist: Playlist, title: String) {
        if (playlist.type == Constants.PLAYLIST_CUSTOM_ID) {
            ApiManager.request(playlist.pid?.let { PlaylistApiServiceImpl.renamePlaylist(it, title) }, object : RequestCallBack<String> {
                override fun success(result: String) {
                    mView.success(1)
                    playlist.name = title
                    EventBus.getDefault().post(MyPlaylistEvent(Constants.PLAYLIST_RENAME, playlist))
                    ToastUtils.show(result)
                }

                override fun error(msg: String) {
                    ToastUtils.show(msg)
                }
            })
        } else {
            doAsync {
                val success = PlaylistLoader.renamePlaylist(playlist, title)
                uiThread {
                    if (success) {
                        mView.success(1)
                        playlist.name = title
                        EventBus.getDefault().post(MyPlaylistEvent(Constants.PLAYLIST_RENAME, playlist))
                        ToastUtils.show("更新成功")
                    }
                }
            }
        }
    }


    companion object {
        private val TAG = "PlaylistDetailPresenter"
    }
}
